File: worker/paidsoon_worker/supabase_environment.py
from __future__ import annotations

import os
import re
from collections.abc import Mapping
from urllib.parse import quote, urlsplit

DEFAULT_POOLER_HOST = "aws-1-ap-southeast-2.pooler.supabase.com"
RUNTIME_PORT = 6543
DATABASE = "postgres"
RUNTIME_QUERY = "pgbouncer=true&connection_limit=1"

_PROJECT_REF_PATTERN = re.compile(r"^[a-z0-9]{20}$")
_POOLER_HOST_PATTERN = re.compile(r"^(?:[a-z0-9-]+\.)+pooler\.supabase\.com$")


class SupabaseConfigError(ValueError):
    def __init__(self, code: str, variable_name: str) -> None:
        self.code = code
        self.variable_name = variable_name
        super().__init__(f"{code}: invalid Supabase configuration for {variable_name}")


def create_database_url(
    project_ref: str | None,
    password: str | None,
    pooler_host: str | None = None,
) -> str:
    if project_ref is None:
        raise SupabaseConfigError(
            "SUPABASE_PROJECT_REF_MISSING", "SUPABASE_PROJECT_REF"
        )
    if not _PROJECT_REF_PATTERN.fullmatch(project_ref):
        raise SupabaseConfigError(
            "SUPABASE_PROJECT_REF_INVALID", "SUPABASE_PROJECT_REF"
        )
    if not password:
        raise SupabaseConfigError(
            "SUPABASE_DB_PASSWORD_MISSING", "SUPABASE_DB_PASSWORD"
        )

    resolved_host = DEFAULT_POOLER_HOST if pooler_host is None else pooler_host
    if not _POOLER_HOST_PATTERN.fullmatch(resolved_host):
        raise SupabaseConfigError(
            "SUPABASE_POOLER_HOST_INVALID", "SUPABASE_DB_POOLER_HOST"
        )

    encoded_password = quote(password, safe="")
    database_url = (
        f"postgresql://postgres.{project_ref}:{encoded_password}@{resolved_host}:"
        f"{RUNTIME_PORT}/{DATABASE}?{RUNTIME_QUERY}"
    )
    parsed = urlsplit(database_url)
    if (
        parsed.scheme != "postgresql"
        or parsed.hostname != resolved_host
        or parsed.port != RUNTIME_PORT
        or parsed.path != f"/{DATABASE}"
        or parsed.query != RUNTIME_QUERY
        or parsed.fragment
    ):
        raise SupabaseConfigError("SUPABASE_DERIVED_URL_INVALID", "DATABASE_URL")
    return database_url


def get_database_url(environ: Mapping[str, str] | None = None) -> str:
    source = os.environ if environ is None else environ
    database_url = create_database_url(
        source.get("SUPABASE_PROJECT_REF"),
        source.get("SUPABASE_DB_PASSWORD"),
        source.get("SUPABASE_DB_POOLER_HOST"),
    )
    legacy_url = source.get("DATABASE_URL")
    if legacy_url is not None and legacy_url != database_url:
        raise SupabaseConfigError("SUPABASE_LEGACY_CONFLICT", "DATABASE_URL")
    return database_url

File: lib/config/supabaseEnvironment.ts
import topology from "../../config/supabase-environment.json"

const PROJECT_REF_PATTERN = /^[a-z0-9]{20}$/
const POOLER_HOST_PATTERN = /^(?:[a-z0-9-]+\.)+pooler\.supabase\.com$/

export type SupabaseConfigErrorCode =
  | "SUPABASE_PROJECT_REF_MISSING"
  | "SUPABASE_PROJECT_REF_INVALID"
  | "SUPABASE_DB_PASSWORD_MISSING"
  | "SUPABASE_DB_PASSWORD_INVALID"
  | "SUPABASE_POOLER_HOST_INVALID"
  | "SUPABASE_DERIVED_URL_INVALID"
  | "SUPABASE_LEGACY_CONFLICT"

export class SupabaseConfigError extends Error {
  readonly code: SupabaseConfigErrorCode
  readonly variableName: string

  constructor(code: SupabaseConfigErrorCode, variableName: string) {
    super(`${code}: invalid Supabase configuration for ${variableName}`)
    this.name = "SupabaseConfigError"
    this.code = code
    this.variableName = variableName
  }
}

export type PublicSupabaseConfig = {
  projectRef: string
  publicUrl: string
}

export type DatabaseSupabaseConfig = PublicSupabaseConfig & {
  poolerHost: string
  databaseUrl: string
  directUrl: string
}

function requireProjectRef(projectRef: string | undefined): string {
  if (projectRef === undefined) {
    throw new SupabaseConfigError(
      "SUPABASE_PROJECT_REF_MISSING",
      "SUPABASE_PROJECT_REF"
    )
  }
  if (!PROJECT_REF_PATTERN.test(projectRef)) {
    throw new SupabaseConfigError(
      "SUPABASE_PROJECT_REF_INVALID",
      "SUPABASE_PROJECT_REF"
    )
  }
  return projectRef
}

function requirePoolerHost(poolerHost: string | undefined): string {
  const resolvedHost = poolerHost === undefined ? topology.defaultPoolerHost : poolerHost
  if (!POOLER_HOST_PATTERN.test(resolvedHost)) {
    throw new SupabaseConfigError(
      "SUPABASE_POOLER_HOST_INVALID",
      "SUPABASE_DB_POOLER_HOST"
    )
  }
  return resolvedHost
}

function encodePassword(password: string | undefined): string {
  if (!password) {
    throw new SupabaseConfigError(
      "SUPABASE_DB_PASSWORD_MISSING",
      "SUPABASE_DB_PASSWORD"
    )
  }

  try {
    return encodeURIComponent(password).replace(
      /[!'()*]/g,
      (character) => `%${character.charCodeAt(0).toString(16).toUpperCase()}`
    )
  } catch {
    throw new SupabaseConfigError(
      "SUPABASE_DB_PASSWORD_INVALID",
      "SUPABASE_DB_PASSWORD"
    )
  }
}

function assertPublicUrl(url: string, projectRef: string): void {
  const parsed = new URL(url)
  if (
    parsed.protocol !== "https:" ||
    parsed.hostname !== `${projectRef}.supabase.co` ||
    parsed.port !== "" ||
    parsed.pathname !== "/" ||
    parsed.search !== "" ||
    parsed.hash !== "" ||
    parsed.username !== "" ||
    parsed.password !== ""
  ) {
    throw new SupabaseConfigError(
      "SUPABASE_DERIVED_URL_INVALID",
      "NEXT_PUBLIC_SUPABASE_URL"
    )
  }
}

function assertDatabaseUrl(options: {
  url: string
  variableName: "DATABASE_URL" | "DIRECT_URL"
  projectRef: string
  encodedPassword: string
  poolerHost: string
  port: number
  runtime: boolean
}): void {
  const parsed = new URL(options.url)
  const expectedQuery = options.runtime
    ? `?pgbouncer=${topology.runtimeQuery.pgbouncer}&connection_limit=${topology.runtimeQuery.connection_limit}`
    : ""

  if (
    parsed.protocol !== "postgresql:" ||
    parsed.username !== `postgres.${options.projectRef}` ||
    parsed.password !== options.encodedPassword ||
    parsed.hostname !== options.poolerHost ||
    parsed.port !== String(options.port) ||
    parsed.pathname !== `/${topology.database}` ||
    parsed.search !== expectedQuery ||
    parsed.hash !== ""
  ) {
    throw new SupabaseConfigError(
      "SUPABASE_DERIVED_URL_INVALID",
      options.variableName
    )
  }
}

export function createPublicSupabaseConfig(
  projectRefInput: string | undefined
): PublicSupabaseConfig {
  const projectRef = requireProjectRef(projectRefInput)
  const publicUrl = `https://${projectRef}.supabase.co`
  assertPublicUrl(publicUrl, projectRef)
  return { projectRef, publicUrl }
}

export function createDatabaseSupabaseConfig(input: {
  projectRef: string | undefined
  password: string | undefined
  poolerHost?: string
}): DatabaseSupabaseConfig {
  const publicConfig = createPublicSupabaseConfig(input.projectRef)
  const poolerHost = requirePoolerHost(input.poolerHost)
  const encodedPassword = encodePassword(input.password)
  const username = `postgres.${publicConfig.projectRef}`
  const databaseUrl =
    `postgresql://${username}:${encodedPassword}@${poolerHost}:` +
    `${topology.runtimePort}/${topology.database}?` +
    `pgbouncer=${topology.runtimeQuery.pgbouncer}&` +
    `connection_limit=${topology.runtimeQuery.connection_limit}`
  const directUrl =
    `postgresql://${username}:${encodedPassword}@${poolerHost}:` +
    `${topology.migrationPort}/${topology.database}`

  assertDatabaseUrl({
    url: databaseUrl,
    variableName: "DATABASE_URL",
    projectRef: publicConfig.projectRef,
    encodedPassword,
    poolerHost,
    port: topology.runtimePort,
    runtime: true,
  })
  assertDatabaseUrl({
    url: directUrl,
    variableName: "DIRECT_URL",
    projectRef: publicConfig.projectRef,
    encodedPassword,
    poolerHost,
    port: topology.migrationPort,
    runtime: false,
  })

  return { ...publicConfig, poolerHost, databaseUrl, directUrl }
}